pkg/apis/pipeline/v1/taskrun_validation.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2012 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/tektoncd/pipeline/pkg/apis/config"
+	"github.com/tektoncd/pipeline/pkg/apis/validate"
+	"github.com/tektoncd/pipeline/pkg/apis/version"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/pkg/apis"
+)
+
+var _ apis.Validatable = (*TaskRun)(nil)
+
+// Validate taskrun
+func (tr *TaskRun) Validate(ctx context.Context) *apis.FieldError {
+	if apis.IsInDelete(ctx) {
+		return nil
+	}
+	errs := validate.ObjectMetadata(tr.GetObjectMeta()).ViaField("metadata")
+	return errs.Also(tr.Spec.Validate(apis.WithinSpec(ctx)).ViaField("spec"))
+}
+
+// Validate taskrun spec
+func (ts *TaskRunSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+	// Must have exactly one of taskRef and taskSpec.
+	if ts.TaskRef == nil && ts.TaskSpec == nil {
+		errs = errs.Also(apis.ErrMissingOneOf("taskRef", "taskSpec"))
+	}
+	if ts.TaskRef != nil && ts.TaskSpec != nil {
+		errs = errs.Also(apis.ErrMultipleOneOf("taskRef", "taskSpec"))
+	}
+	// Validate TaskRef if it's present.
+	if ts.TaskRef != nil {
+		errs = errs.Also(ts.TaskRef.Validate(ctx).ViaField("taskRef"))
+	}
+	// Validate TaskSpec if it's present.
+	if ts.TaskSpec != nil {
+		errs = errs.Also(ts.TaskSpec.Validate(ctx).ViaField("taskSpec"))
+	}
+
+	errs = errs.Also(ValidateParameters(ctx, ts.Params).ViaField("params"))
+	errs = errs.Also(ValidateWorkspaceBindings(ctx, ts.Workspaces).ViaField("workspaces"))
+	if ts.Debug != nil {
```

Comment:
Shall I sync with 5291 here for the [Resources related code](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/taskrun_validation.go/#L68)

Reply:
I'm not exactly sure what you mean here-- we shouldn't have resources in v1 so no need to validate them

Reply:
Sorry this was mistaken with some other changes when I looked at the change log.